### PR TITLE
core workflow: simplify file search results display

### LINKED ISSUE #38182: Clean up unnecessary icons
We show some icons that feel redundant, like a file icon when viewing what is obviously a match in the file contents:

Image

We also show both an “S” icon identifying a result as a symbol type but then also include an icon for the type of symbol it is:

Image

Image

Other feedback from @novoselrok and @fkling:
@novoselrok:
>I find our symbol icons, in general, very hard to parse. I prefer GitHub's way of having the symbol initials with different colors as the icon. E.g. `fn` for functions, `class` for classes, etc

@fkling 
>Same. The icons are not self-explanatory so you'd have to learn them. Using (some form of) type name would make it easier for people to "get into it". I also like how GitHub prefixes symbols with different colored type names.

### LINKED ISSUE #38566: Simplify file content result display
Currently, we provide a pretty intense user experience when rendering files that have many matches in them, depending on the user’s set amount of context lines.

Even with context lines set at 2, with 4 matches in a file it feels like a lot for the default.

Image

We’d like to experiment with something similar to what GitHub does with a fixed result size, no matter how many matches are in the file.

After talking with @novoselrok we may be a little limited in what we can do here, or at least constrained in our options by the existing results.

## Proposal

Comment from @fkling:
>We could also consider to not show contextual lines.
>If we can make the blob view fast and show it directly next to search results (not as an overlay) then this might be good enough.
>Alternatively we could add some way to reveal contextual lines before/after.
And of course the user could configure to always show a minimal number of contextual lines (just like it is now).

### PR DESCRIPTION
**Pull request by novoselrok**
Fixes #38566
Fixes #38182

Changes (all contained behind the core workflow temporary setting):
* Add a button for expanding/collapsing matches below the file search result
* Reduce the number of line groups shown by default
* Hide the result type icons

## Test plan

* Disable the core workflow improvements (Simple UI toggle in user settings dropdown)
  * Search results should work the same as before
* Enable Simple UI
  * Run a search
  * You should be able to expand files with a lot of matches
  * No result type icons should appear next to the repo name/file name

### CODE REVIEW

File: client/search-ui/src/components/FileSearchResult.tsx [hunk 1]
```diff
@@ -230,87 +255,53 @@ export const FileSearchResult: React.FunctionComponent<React.PropsWithChildren<P
             { limitedGrouped: [] as MatchGroup[], limitedMatchCount: 0 }
         )
 
+        const collapsedChildren = <FileMatchChildren {...props} result={result} grouped={limitedGrouped} />
+        const expandedChildren = <FileMatchChildren {...props} result={result} grouped={grouped} />
+
         if (props.showAllMatches) {
             containerProps = {
-                index: props.index,
```
Comment (novoselrok) [inline client/search-ui/src/components/FileSearchResult.tsx:235]: The duplicated props were getting annoying, so I decided to clean them up.

**Comment by olafurpg** [comment]
I like the idea of placing the expand button at the bottom of the file result. One concern I have is that you need to scroll down to collapse again. Is it possible to have a collapse button somewhere in the gutter where the expand button is to make it easier to quickly click on expand/collapse. Feel free to ignore, just something I thought about

CleanShot 2022-07-14 at 08 25 05

**Comment by olafurpg** [comment]
Alternatively, maybe there's a way to keep the row with the "expand" button between the hunks. Users don't need to drag/select text across different hunks so it's maybe fine to interleave it with a collapse button 🤔

**Comment by novoselrok** [comment]
> I like the idea of placing the expand button at the bottom of the file result. One concern I have is that you need to scroll down to collapse again. Is it possible to have a collapse button somewhere in the gutter where the expand button is to make it easier to quickly click on expand/collapse. Feel free to ignore, just something I thought about

That's a good point. I'll let @benvenker and @almeidapaulooliveira chime in. If it's simple enough I can add it to this PR, otherwise, we can explore this separately.

**Comment by benvenker** [comment]
@olafurpg are you thinking it would stick just while you're scrolling through that result and disappear when the result disappeared from view?

**Comment by benvenker** [comment]
> 

Makes sense to me. @novoselrok if you want to flip them when clicked I'm good to go ahead and do that!

**Review by benvenker — APPROVED** [review]

**Comment by novoselrok** [comment]
> Makes sense to me. @novoselrok if you want to flip them when clicked I'm good to go ahead and do that!

As discussed, the current search results component does not allow embedding buttons in the middle of hunks (not without heavy hacking at least). We can spin this off into a separate issue.

**Comment by benvenker** [comment]
@novoselrok @almeidapaulooliveira made some small updates to the result component. Figma file is [here](https://www.figma.com/file/8MkIcpDtjv6tqgxZFcEQDt/Core-workflow-punch-list?node-id=176%3A18700) and prototype is [here](https://www.figma.com/proto/8MkIcpDtjv6tqgxZFcEQDt/Core-workflow-punch-list?page-id=0%3A1&node-id=176%3A18700&viewport=585749%2C385342%2C26.68&scaling=min-zoom&starting-point-node-id=176%3A19861).

The things that have changed are:
- The file header link color
- There is an outline on hover of the search result component (not an individual match within it)

One question that I'm blanking on your answer to: Is it possible to just show the first match in a result? What I'm getting at is visual consistency for all un-expanded search results. So in the screenshot below, for example, I just want the first match chunk in the third result. Does that make sense?
image

I have a memory that we talked about this yesterday morning and you said it was complicated because of our search result structure, but I might be thinking of something else.

**Comment by novoselrok** [comment]
@benvenker I added the changes that you linked. I chose the closest colors I could find in our colors.scss file.

> One question that I'm blanking on your answer to: Is it possible to just show the first match in a result?

It's tricky. The best we could do right now is to show only the first *group* of lines which can still contain multiple matches. I think that's a fair approximation of the designs.

**Review by benvenker — APPROVED** [review]

File: client/search-ui/src/components/RepoSearchResult.tsx [hunk 2]
```diff
@@ -107,7 +110,7 @@ export const RepoSearchResult: React.FunctionComponent<RepoSearchResultProps> =
     return (
         <ResultContainer
             index={index}
-            icon={SourceRepositoryIcon}
+            icon={!coreWorkflowImprovementsEnabled ? SourceRepositoryIcon : undefined}
```
Comment (fkling) [inline client/search-ui/src/components/RepoSearchResult.tsx:113]: I assume we still show icons for other kinds of search results? Because if not then we could just hide the icon in CSS via the global class instead of conditionally setting it.

**Review by fkling — APPROVED** [review]

**Comment by fkling** [comment]
Two additional ideas but they could be addressed in another diff: Make the result heading and the collapse button sticky so that they don't scroll out of view for a given search result block.

**Comment by limitedmage** [comment]
@novoselrok I can make the collapse button stick to the bottom of the screen (since I'm now painfully familiar with `position:sticky` and its caveats), let me know if you'd like me to do that as follow-up to this PR.

**Comment by almeidapaulooliveira** [comment]
@novoselrok - That's a different color (lighter) on the line codes too. @benvenker forget to mention.

**Comment by novoselrok** [comment]
> @novoselrok I can make the collapse button stick to the bottom of the screen (since I'm now painfully familiar with `position:sticky` and its caveats), let me know if you'd like me to do that as follow-up to this PR.

@limitedmage Yep, go ahead in a follow-up 🙂
